drivers/sdhc/sam_hsmci.c

```diff
@@ -0,0 +1,537 @@
+/*
+ * Copyright 2023 Nikhef
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam_hsmci
+
+#include <zephyr/drivers/sdhc.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/atmel_sam_pmc.h>
+
+#ifndef HSMCI_MR_PDCMODE
+#error "Sorry, only SAM4E/SAM4S are supported at the moment"
+#endif
+
+LOG_MODULE_REGISTER(hsmci, CONFIG_SDHC_LOG_LEVEL);
+
+#define _HSMCI_DEFAULT_TIMEOUT 5000
+#define _HSMCI_MAX_FREQ        (SOC_ATMEL_SAM_MCK_FREQ_HZ >> 1)
+#define _HSMCI_MIN_FREQ        (_HSMCI_MAX_FREQ / 0x200) // only valid when MCI has ODD bit
+#define _HSMCI_PWRSAVE         7
+
+#define _HSMCI_SR_ERR                                                                              \
+	(HSMCI_SR_RINDE | HSMCI_SR_RDIRE | HSMCI_SR_RCRCE | HSMCI_SR_RENDE | HSMCI_SR_RTOE |       \
+	 HSMCI_SR_DCRCE | HSMCI_SR_DTOE | HSMCI_SR_CSTOE | HSMCI_SR_OVRE | HSMCI_SR_UNRE)
+
+static const uint8_t _resp2size[] = {
+	[SD_RSP_TYPE_NONE] = HSMCI_CMDR_RSPTYP_NORESP,
+	[SD_RSP_TYPE_R1] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R1b] = HSMCI_CMDR_RSPTYP_R1B,
+	[SD_RSP_TYPE_R2] = HSMCI_CMDR_RSPTYP_136_BIT,
+	[SD_RSP_TYPE_R3] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R4] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R5] = 0 /*?*/,
+	[SD_RSP_TYPE_R5b] = 0 /*?*/,
+	[SD_RSP_TYPE_R6] = HSMCI_CMDR_RSPTYP_48_BIT,
+	[SD_RSP_TYPE_R7] = HSMCI_CMDR_RSPTYP_48_BIT,
+};
+
+// timeout multiplier shift (actual value is 1 << _mul_shift[*])
+static const uint8_t _mul_shift[] = {0, 4, 7, 8, 10, 12, 16, 20};
+static const uint8_t _mul_shift_size = 8;
+
+struct sam_hsmci_config {
+	Hsmci *base;
+	const struct atmel_sam_pmc_config clock_cfg;
+	const struct pinctrl_dev_config *pincfg;
+	struct gpio_dt_spec carrier_detect;
+};
+
+struct sam_hsmci_data {
+	bool open_drain;
+	uint8_t cmd_in_progress;
+	struct k_mutex mtx;
+};
+
+static int sam_hsmci_reset(const struct device *dev)
+{
+	LOG_DBG("sam_hsmci_reset()");
```

Comment:
Minor nit here: why not use `LOG_DBG("%s()", __func__)`?

Also, please move this LOG statement after variable declarations